receiver.go

```diff
@@ -55,29 +42,26 @@ type receiveHandler struct {
 	messageHandlers map[MethodType]messageHandler
 }
 
-func newReceiveHandler(enc encoder, dec byteReadingDecoder, rmCallCh chan callRetrieval, l LogInterface, wef WrapErrorFunc) *receiveHandler {
+func newReceiveHandler(enc encoder, protHandler *protocolHandler, l LogInterface) *receiveHandler {
 	r := &receiveHandler{
-		writer:    enc,
-		reader:    dec,
-		protocols: make(map[string]Protocol),
-		tasks:     make(map[int]context.CancelFunc),
-		listeners: make(map[chan error]struct{}),
-		rmCallCh:  rmCallCh,
-		stopCh:    make(chan struct{}),
-		closedCh:  make(chan struct{}),
+		writer:      enc,
+		protHandler: protHandler,
+		tasks:       make(map[int]context.CancelFunc),
+		listeners:   make(map[chan<- error]struct{}),
+		stopCh:      make(chan struct{}),
+		closedCh:    make(chan struct{}),
 
 		taskBeginCh:  make(chan *task),
 		taskCancelCh: make(chan seqNumber),
 		taskEndCh:    make(chan seqNumber),
 
-		log:           l,
-		wrapErrorFunc: wef,
+		log: l,
 	}
 	r.messageHandlers = map[MethodType]messageHandler{
-		MethodNotify:   {dispatchFunc: r.receiveNotify, messageLength: 3},
-		MethodCall:     {dispatchFunc: r.receiveCall, messageLength: 4},
-		MethodResponse: {dispatchFunc: r.receiveResponse, messageLength: 4},
-		MethodCancel:   {dispatchFunc: r.receiveCancel, messageLength: 3},
+		MethodNotify:   r.receiveNotify,
```

Comment:
is it necessary to have this map now? You could just do a switch/case below in Receive.